platform/linux-generic/odp_ipsec_sad.c

```diff
@@ -423,6 +423,39 @@ static uint32_t esp_block_len_to_mask(uint32_t block_len)
 	return block_len - 1;
 }
 
+/* AR window management initialization */
+static int ipsec_antireplay_init(ipsec_sa_t *ipsec_sa,
+				 const odp_ipsec_sa_param_t *param)
+{
+	uint16_t num_bkts = 0;
+
+	if (param->inbound.antireplay_ws > IPSEC_AR_WIN_SIZE_MAX) {
+		ODP_ERR("AR_WS[%u] not supported\n", param->inbound.antireplay_ws);
+		return -1;
+	}
+
+	ipsec_sa->antireplay = (param->inbound.antireplay_ws != 0);
+	if (!ipsec_sa->antireplay)
+		return 0;
+
+	ipsec_sa->hot.in.win_size = ROUNDUP_POWER2_U32(param->inbound.antireplay_ws);
```

Comment:
What is the purpose of this rounding? As I commented about an earlier version of this PR, as far as I can see everything would work fine without this rounding. However, with suitable rounding the number of buckets could be made a power of two so that figuring out the right bucket would not need an integer division but just a shift.

Reply:
> What is the purpose of this rounding?

As part of the NOKIA requirement, only the AR window sizes 32, 64, 128, 256, 512, 1024, 2048 and 4096 are supposed to be implemented, hence rounding the AR window sizes to the next power of 2 and figuring out the minimum required  bit buckets from the rounded window sizes. 
 
>However, with suitable rounding the number of buckets could be made a power of two so that figuring out the right bucket would not need an integer division but just a shift.

Yes agreed, integer division would be avoided, but unfortunately could not make number of buckets as power of two, because for each WS one extra bucket is required to avoid the loss of very first bucket bits inside the window(to avoid history loss) during the bucket recycling process. 

===================
WS       |   Minimum buckets
==== ==============
64        |        2 
128      |        3 
256      |        5 
512      |        9 
1026    |       17
2048    |       33
4096    |       65
==================

One solution to solve the integer division issue is to make the effective window size implemented as application WS minus 64, this way the buckets can be rounded to the power of two. With this solution 
============================================
Application WS    |    Implemented WS     |       buckets
============================================
64                         |        64                        |           2     
128                       |        64                        |           2 
256                       |      192   ( 256 -64)     |           4
512                       |       448  (512 -64)      |           8
1024                     |       960  (1024 -64)    |          16 
2048                     |      1984 ( 2048 -64)   |          32 
4096                     |       4032( 4086 -64)   |          64
=============================================


> as far as I can see everything would work fine without this rounding.

Yes, we agree that this implementation would work for all the window sizes till 4096. So if you like to make it generic to work for all the window sizes, please confirm so that we will work on any required changes.

Reply:
> As part of the NOKIA requirement, only the AR window sizes 32, 64, 128, 256, 512, 1024, 2048 and 4096 are supposed to be implemented, hence rounding the AR window sizes to the next power of 2

Discussion of possible requirements agreed between you and Nokia does not belong here.

That said, if there is a requirement to support certain window sizes, it does not mean that other sizes must not be supported. Linux-generic ODP can support things that are not included in some random requirement list somewhere. In this case there is no point having additional code (the rounding stuff) whose only purpose is to limit the functionality if the implementation does not take any advantage of the limitation and would work equally well without the limitation.

Reply:
> > As part of the NOKIA requirement, only the AR window sizes 32, 64, 128, 256, 512, 1024, 2048 and 4096 are supposed to be implemented, hence rounding the AR window sizes to the next power of 2
> 
> Discussion of possible requirements agreed between you and Nokia does not belong here.
> 
> That said, if there is a requirement to support certain window sizes, it does not mean that other sizes must not be supported. Linux-generic ODP can support things that are not included in some random requirement list somewhere. In this case there is no point having additional code (the rounding stuff) whose only purpose is to limit the functionality if the implementation does not take any advantage of the limitation and would work equally well without the limitation.

Okay, will remove the window size rounding, this means that
 

- All the application window sizes ranging from 33 to 4096 will be considered as they are by the implementation, there will be no adjustments done by the implementation.
- All the application window sizes ranging from 1 to 32 are adjusted to window size 32 by the implementation.